ssm_jax/ekf/inference.py

```diff
@@ -0,0 +1,195 @@
+    filtered state estimates.
+
+    Args:
+        params: an NLGSSMParams instance (or object with the same fields)
+        emissions (T,D_hid): array of observations.
+        inputs (T,D_in): array of inputs.
+
+    Returns:
+        filtered_posterior: ESSMPosterior instance containing,
+            marginal_log_lik 
+            filtered_means (T, D_hid)
+            filtered_covariances (T, D_hid, D_hid)
+    """
+    num_timesteps = len(emissions)
+    # Dynamics and emission functions and their Jacobians
+    f, h = params.dynamics_function, params.emission_function
+    F, H = jacfwd(f), jacfwd(h)
+    # If no input, add dummy input to functions
+    if inputs is None:
+        inputs = jnp.zeros((num_timesteps,))
+        process_fn = lambda fn: (lambda x, u: fn(x))
+        f, h, F, H = (process_fn(fn) for fn in (f, h, F, H))
+
+    def _step(carry, t):
+        ll, pred_mean, pred_cov = carry
+
+        # Get parameters and inputs for time index t
+        Q = _get_params(params.dynamics_covariance, 2, t)
+        R = _get_params(params.emission_covariance, 2, t)
+        u = inputs[t]
+        y = emissions[t]
+
+        # Update the log likelihood
+        H_x = H(pred_mean, u)
+        ll += MVN(h(pred_mean, u), H_x @ pred_cov @ H_x.T + R).log_prob(y)
+
+        # Condition on this emission
+        filtered_mean, filtered_cov = _condition_on(
+            pred_mean, pred_cov, h, H, R, u, y)
+
+        # Predict the next state
+        pred_mean, pred_cov = _predict(
+            filtered_mean, filtered_cov, f, F, Q, u)
+
+        return (ll, pred_mean, pred_cov), (filtered_mean, filtered_cov)
+
+    # Run the extended Kalman filter
+    carry = (0., params.initial_mean, params.initial_covariance)
+    (ll, _, _), (filtered_means, filtered_covs) = lax.scan(
+        _step, carry, jnp.arange(num_timesteps))
+    return NLGSSMPosterior(marginal_loglik=ll,
+                           filtered_means=filtered_means,
+                           filtered_covariances=filtered_covs)
+
+
+def extended_kalman_smoother(params, emissions, inputs=None):
+    """Run an extended Kalman (RTS) smoother.
+
+    Args:
+        params: an NLGSSMParams instance (or object with the same fields)
+        emissions (T,D_hid): array of observations.
+        inputs (T,D_in): array of inputs.
+
+    Returns:
+        nlgssm_posterior: ESSMPosterior instance containing properties of
+            filtered and smoothed posterior distributions.
+    """
+    num_timesteps = len(emissions)
+    
+    # Run the extended Kalman filter
+    ekf_posterior = extended_kalman_filter(params, emissions, inputs)
+    ll, filtered_means, filtered_covs, *_ = ekf_posterior.to_tuple()
+    
+    # Dynamics and emission functions and their Jacobians
+    f, h = params.dynamics_function, params.emission_function
+    F, H = jacfwd(f), jacfwd(h)
+    # If no input, add dummy input to functions
+    if inputs is None:
```

Comment:
Can this be factored out into a helper function, similar to `_get_params`?

Reply:
Good idea! Thank you.